pkg/apiserver/metrics/deployments.go

```diff
@@ -0,0 +1,168 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"k8s.io/kubernetes/pkg/api/rest"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"golang.org/x/net/context"
+)
+
+var (
+	variableLabels = []string{"name", "namespace"}
+
+	deploymentStatusObservedGenerationDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_observed_generation",
+		"The observed generation of this deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_replicas",
+		"The number of replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusUpdatedReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_updated_replicas",
+		"The number of updated replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusAvailableReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_available_replicas",
+		"The number of available replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusUnavailableReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_unavailable_replicas",
+		"The number of unavailable replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+)
+
+// DeploymentCollector collects various metrics of deployment objects.
+type DeploymentCollector struct {
+	Storage rest.StandardStorage
+}
+
+// Returns a new instance of DeploymentCollector.
+func NewDeploymentsCollector(s rest.StandardStorage) *DeploymentCollector {
+	return &DeploymentCollector{
+		Storage: s,
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *DeploymentCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- deploymentStatusObservedGenerationDesc
+	ch <- deploymentStatusReplicasDesc
+	ch <- deploymentStatusUpdatedReplicasDesc
+	ch <- deploymentStatusAvailableReplicasDesc
+	ch <- deploymentStatusUnavailableReplicasDesc
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *DeploymentCollector) Collect(ch chan<- prometheus.Metric) {
+	var err error
+	deployments, err := c.Storage.List(context.Background(), nil)
+	if err != nil {
+		glog.Errorf("Error while collecting deployment metrics: %s", err.Error())
+	}
+
+	for _, d := range deployments.(*extensions.DeploymentList).Items {
+		err = addMetrics(ch, d)
+		if err != nil {
+			glog.Errorf("Error while collecting deployment metrics: %s", err.Error())
+		}
+	}
+}
+
+// addMetrics adds Metrics for a single deployment to the specified channel
+func addMetrics(ch chan<- prometheus.Metric, d extensions.Deployment) error {
```

Comment:
Interesting. Could you point us to where exactly you mean? We are not too deep into the Kubernetes code base yet.

Generally speaking, this instrumentation is about the apiserver's fundamental functionality. It is inconsistent to have some basics natively instrumented and the actually useful stuff somewhere outside. Metrics should be treated like logging in that regard.

Tracking it outside would mean querying the API directly for way more information than we need and having to extract it from the response. It also forever couples instrumentation to what is accessible via the API, which is not always congruent.
Cost only occurs when the `/metrics` endpoint is actually being scraped. So if we want to monitoring, pushing the instrumentation outside just means imposing more load on our system.

In case of a cost concern, we could discuss making different sets of metrics available differently. Either by providing multiple paths, e.g. `/metrics/deployments`, `/metrics/replica_sets`, etc., or via query parameters enabling/disabling different collectors, e.g. `/metrics?collector=deployments` etc.
This would also allow for different sampling intervals for different metric sets.

@brian-brazil and @beorn7 for opinions and comments on metric library efficiency.